### PR TITLE
Scale retro terminal scan lines

### LINKED ISSUE #4362: Retro terminal effect prodces artifacts on the right screen edge
# Environment

```none
Windows build number: 10.0.19551.1005
Windows Terminal version: 0.8.10091.0

Any other software?
Confirmed in the PowerShell 7.0.0-rc2 and cmd shells, but should affect all.
```

# Steps to reproduce

Set your PowerShell Core (7) profile like or similar to this:

```json
"acrylicOpacity":  0.84,
"fontFace":  "Cascadia Code",
"fontSize":  11,
"hidden":  false,
"name":  "PowerShell 7",
"experimental.retroTerminalEffect": true,
"source":  "Windows.Terminal.PowershellCore",
"padding":  "25, 25, 25, 25",
"useAcrylic":  true
```

# Expected behavior

No glowy artifacts on any screen border/edge.

# Actual behavior

image

When retro terminal effects AND screenbuffer padding is enabled, text-effect artifacts are visible on the right edge in many lines (not all). This will also happen without padding but it's visually less obvious.

### PR DESCRIPTION
**Pull request by dotpaul**
## Summary of the Pull Request
- Scale the retro terminal effects (#3468) scan lines with the screen's DPI.
- Remove artifacts from sampling wrap around.

Before & after, with my display scale set to 350%:
Scaling scan lines

Before & after showing artifact removal, with my display scale set to 100%, and image enlarged to 400%:
Sampling artifacts annotated

## PR Checklist
* [x] Closes #4362
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

Adds a constant buffer, which could be used for other settings for the retro terminal pixel shader.

I haven't touched C++ in over a decade before this change, and this is the first time I've played with DirectX, so please assume my code isn't exactly best practice. 🙂

## Validation Steps Performed

- Changed display scale with experimental.retroTerminalEffect enabled, enjoyed scan lines on high resolution monitors.
- Enabled experimental.retroTerminalEffect, turned the setting off, changed display scale. Retro tabs still scale scan lines.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
YES THANK YOU

**Comment by DHowett-MSFT** [comment]
You can add `Closes #4362` for the screen edge issue

**Comment by dotpaul** [comment]
You're welcome! Tagged the screen edge issue.

File: src/renderer/dx/DxRenderer.cpp [hunk 1]
```diff
@@ -285,12 +285,24 @@ HRESULT DxEngine::_SetupTerminalEffects()
 
     RETURN_IF_FAILED(_d3dDevice->CreateBuffer(&bd, &InitData, &_screenQuadVertexBuffer));
 
+    D3D11_BUFFER_DESC pixelShaderSettingsBufferDesc{};
+    pixelShaderSettingsBufferDesc.Usage = D3D11_USAGE_DEFAULT;
+    pixelShaderSettingsBufferDesc.ByteWidth = sizeof(_pixelShaderSettings) + (16 - sizeof(_pixelShaderSettings) % 16) % 16;
```
Comment (zadjii-msft) [inline src/renderer/dx/DxRenderer.cpp]: Okay I don't know DX that well but _what the heck is going on here_

File: src/renderer/dx/DxRenderer.cpp [hunk 2]
```diff
@@ -285,12 +285,24 @@ HRESULT DxEngine::_SetupTerminalEffects()
...
+
+    RETURN_IF_FAILED(_d3dDevice->CreateBuffer(&pixelShaderSettingsBufferDesc, &pixelShaderSettingsInitData, &_pixelShaderSettingsBuffer));
+
     // Sampler state is needed to use texture as input to shader.
     D3D11_SAMPLER_DESC samplerDesc{};
     samplerDesc.Filter = D3D11_FILTER_MIN_MAG_MIP_LINEAR;
-    samplerDesc.AddressU = D3D11_TEXTURE_ADDRESS_WRAP;
-    samplerDesc.AddressV = D3D11_TEXTURE_ADDRESS_WRAP;
-    samplerDesc.AddressW = D3D11_TEXTURE_ADDRESS_WRAP;
+    samplerDesc.AddressU = D3D11_TEXTURE_ADDRESS_CLAMP;
```
Comment (zadjii-msft) [inline src/renderer/dx/DxRenderer.cpp]: THANK YOU

**Comment by zadjii-msft** [comment]
@msftbot make sure @miniksa signs off on this

File: src/renderer/dx/DxRenderer.cpp (hunk 1 above)
Comment (DHowett-MSFT) [reply src/renderer/dx/DxRenderer.cpp]: I believe this is "nearest larger multiple of 16"

```
libra  ~  %   $n=9
libra  ~  %   $n + (16 - $n % 16) % 16
16
libra  ~  %   $n=17
libra  ~  %   $n + (16 - $n % 16) % 16
32
```

Comment (DHowett-MSFT) [reply src/renderer/dx/DxRenderer.cpp]: (that "multiples of 16" thing is documented in the docs for `CreateBuffer` for when you pass `D3D11_BIND_CONSTANT_BUFFER`)

File: src/renderer/dx/DxRenderer.cpp [hunk 3]
```diff
@@ -1587,6 +1616,16 @@ CATCH_RETURN()
...
     RETURN_IF_FAILED(InvalidateAll());
 
+    if (_retroTerminalEffects && _d3dDeviceContext && _pixelShaderSettingsBuffer)
+    {
+        _ComputePixelShaderSettings();
+        try
+        {
+            _d3dDeviceContext->UpdateSubresource(_pixelShaderSettingsBuffer.Get(), 0, NULL, &_pixelShaderSettings, 0, 0);
+        }
+        CATCH_RETURN();
```
Comment (DHowett-MSFT) [inline src/renderer/dx/DxRenderer.cpp]: Sorry, I don't know if my previous comment actually got saved on the backend b/c of the github 500 errors.

I think this should be `RETURN_IF_FAILED(_d3dContext->UpdateSubresource(x, y, z))`. UpdateSubresource won't throw an exception b/c it's on the other side of a COM barrier, it'll return an HRESULT (which right now we're discarding)

File: src/renderer/dx/DxRenderer.cpp (hunk 1 above)
Comment (dotpaul) [reply src/renderer/dx/DxRenderer.cpp]: Right. Thinking about this more, I want to pad the struct instead. As it is now, I would assume DirectX is reading beyond the struct's memory when initializing / updating the constant buffer, which would make me feel bad.

File: src/renderer/dx/DxRenderer.cpp (hunk 3 above)
Comment (dotpaul) [reply src/renderer/dx/DxRenderer.cpp]: `UpdateSubresource` returns a void. Should I do something else instead?

I added the `try` cuz the static analysis build was giving this error: `Error C26447: The function is declared 'noexcept' but calls function 'UpdateSubresource()' which may throw exceptions (f.6).`

File: src/renderer/dx/DxRenderer.cpp (hunk 1 above)
Comment (dotpaul) [reply src/renderer/dx/DxRenderer.cpp]: Thanks for pointing that out, @zadjii-msft !

File: src/renderer/dx/DxRenderer.cpp (hunk 3 above)
Comment (DHowett-MSFT) [reply src/renderer/dx/DxRenderer.cpp]: oh, that's interesting. meh. stupid static analyzer can't be taught that COM things don't throw exceptions. Fine, this is fine.

**Comment by DHowett-MSFT** [comment]
@msftbot make sure @miniksa signs off

File: src/renderer/dx/ScreenPixelShader.h [hunk 4]
```diff
@@ -74,7 +78,7 @@ float4 main(float4 pos : SV_POSITION, float2 tex : TEXCOORD) : SV_TARGET
 
     // TODO:GH#3930 Make these configurable in some way.
```
Comment (miniksa) [inline src/renderer/dx/ScreenPixelShader.h]: Well this TODO is going to be a lot easier now that you piped through the settings buffer.

**Comment by DHowett-MSFT** [comment]
I pushed a merge to your branch so we could land it :smile:

**Comment by dotpaul** [comment]
> I pushed a merge to your branch so we could land it 😄

Ah, thank you @DHowett-MSFT !

**Pull request merged**